ddtrace/contrib/internal/asgi/middleware.py

```diff
@@ -241,9 +259,288 @@ async def __call__(self, scope, receive, send):
             )
             tags = _extract_versions_from_scope(scope, self.integration_config)
             span.set_tags(tags)
+            global_root_span = span._local_root or span
+
+            @wraps(receive)
+            async def wrapped_receive():
+                """
+                websocket.message.type
+                websocket.message.length
+                websocket.message.frames
+                """
+
+                if not self.integration_config._trace_asgi_websocket_messages:
+                    return await receive()
+
+                # Create the span when the handler is invoked (when receive() is called)
+                # This measures the time from receive() call to message processing completion
+                recv_span = self.tracer.start_span(
+                    name="websocket.receive",
+                    service=span.service,
+                    resource=f"websocket {scope.get('path', '')}",
+                    span_type="websocket",
+                    child_of=span if not self.integration_config._websocket_messages_separate_traces else None,
+                    activate=True,
+                )
+                try:
+                    message = await receive()
+
+                    if scope["type"] == "websocket" and message["type"] == "websocket.receive":
+                        # Keep the current receive span open to measure processing time
+                        # It will be finished after the application processes the message
+
+                        # end current receive span before
+                        current_receive_span = scope.get("datadog", {}).get("current_receive_span")
+                        if current_receive_span:
+                            current_receive_span.finish()
+                            scope["datadog"].pop("current_receive_span", None)
+
+                        core.dispatch("asgi.websocket.receive", (message,))
+                        recv_span.set_tag_str(COMPONENT, self.integration_config.integration_name)
+                        recv_span.set_tag_str(SPAN_KIND, SpanKind.CONSUMER)
+
+                        if "text" in message:
+                            recv_span.set_tag_str("websocket.message.type", "text")
+                            recv_span.set_metric("websocket.message.length", len(message["text"].encode("utf-8")))
+                        elif "binary" in message:
+                            recv_span.set_tag_str("websocket.message.type", "binary")
+                            recv_span.set_metric("websocket.message.length", len(message["bytes"]))
+
+                        recv_span.set_metric("websocket.message.frames", 1)
+
+                        recv_span.set_link(
+                            trace_id=span.trace_id, span_id=span.span_id, attributes={"dd.kind": "executed_by"}
```

Comment:
Can we use a constant for these attributes. It seems like something that is used in a few integrrations.